Package.json validation tests

teallarson opened the pull request:
## Change description

- Adds schema validation tests for our Package.json files
- Checks for version matches across the repo 
- Checks file contents for items in our "bad words" list

Went with invoking the version checker and spelling checker within child process inside the jest test as the scripts in place did a more efficient job of the tasks then a JavaScript solution would in my opinion.

Also removed the CI jobs that used to execute those scripts, and added one to run the new suite.

## Checklists

### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

Commit:
initial commit

Commit:
package json changes

Commit:
pnpm things

Commit:
tests completed, scripts moved/erased

Commit:
more moving things

Commit:
ci configurations

Commit:
change ci_generate

Commit:
ci configuration

Commit:
try

Commit:
try

grouparoo-bot committed:
Updating things automatically

teallarson commented:
the flaky test here in models is fixed by #2943 (missing awaits)

Commit:
remove moved scripts

evantahler commented on jest.setup.js:
```diff
@@ -0,0 +1,4 @@
+require("@grouparoo/spec-helper/dist/lib/environment");
+
+// tell `@grouparoo/spec-helper` to load everything from `src` in core, so we can test without re-compiling
+process.env.GROUPAROO_SPEC_HELPER_LOAD_PATH = "src";
```
This probably isn't required at the top-level of the monorepo - this is for running the grouparoo app, which we don't need here

evantahler commented on package.json:
```diff
@@ -13,19 +13,23 @@
   },
   "devDependencies": {
+    "@grouparoo/core": "0.8.1-alpha.0",
+    "@grouparoo/spec-helper": "0.8.1-alpha.0",
```
Having the top of the monorepo depend on something within it might get wonky.  I don't think this is needed if `jest.setup` is removed

evantahler commented on tools/merger/data/ci/root/job.yml.template:
```diff
@@ -0,0 +1,29 @@
+          username: $DOCKERHUB_USERNAME
+          password: $DOCKERHUB_PASSWORD
+        environment:
+          POSTGRES_PASSWORD: password
```
I don't think the root tests need redis or postgres to run

evantahler replied:
Bump! I don't think you need the docker images other than {{{docker_image}}}

evantahler commented on tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+    "target": "es2019",
+    "esModuleInterop": true,
+    "experimentalDecorators": true,
+    "emitDecoratorMetadata": true
```
These last 2 probably aren't needed, the rest looks good!

evantahler commented on tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+    "experimentalDecorators": true,
+    "emitDecoratorMetadata": true
+  },
+  "include": ["./src/**/*"],
```
There is no `src` in the root project, so maybe this isn't needed?

evantahler commented on jest.config.js:
```diff
@@ -0,0 +1,18 @@
+const { helper } = require("@grouparoo/spec-helper");
```
This can probably be skipped, and just set the testTimeout manually - when we don't need this dependency

evantahler commented on jest.config.js:
```diff
@@ -0,0 +1,18 @@
+  transform: {
+    "^.+\\.ts?$": "ts-jest",
+  },
+  testPathIgnorePatterns: [
```
I don't think any of these directories exist in the root project

evantahler commented on __tests__/packages.ts:
```diff
@@ -0,0 +1,119 @@
+          if (error || stderr) {
+            reject(error || stderr);
+          }
+          res();
```
without a return statement, this will both resolve and reject this promise.  either `return reject(...)` or do an if/else

evantahler replied:
What does it look like when this promise throws?  The cli usage of syncpack did a good job of showing us which packages are mismatched... does that come through in the jest error?

krishnaglick replied:
`done(error || stderr)` pattern worth considering like I mentioned above

teallarson replied:
I think what we actually what here is `return reject(stdout)` since that's where the useful info is sent.

evantahler commented on __tests__/schemaGeneric.json:
```diff
@@ -0,0 +1,244 @@
+          "$id": "#/properties/publishConfig/properties/access",
+          "type": "string",
+          "examples": ["public"]
+        }
```
This is confirming that `publishConfig.access` is a string in the package.json, but it might not be `"public"` - can we confirm that it is literally "public" (except for ui-enterprise)?

teallarson replied:
Yes.  I did set that on the plugin-specific schema as I assumed we only guaranteed that the plugins would have `publishConfig`.  Are you saying we want that on every package.json except `ui-enterprise`?

evantahler replied:
Sorry about that! I'm more interested in confirming that most packages have a literal `publishConig.access = 'public'` rather than any string, which I think is what this assertion does.  I would prefer a validation/test that checks for that value directly vs that any string is present.  

The nuance is that this is true for all packages except ui-enterprise.

evantahler commented on __tests__/schemaGeneric.json:
```diff
@@ -0,0 +1,244 @@
+      "properties": {
+        "node": {
+          "$id": "#/properties/engines/properties/node",
+          "type": "string",
```
Same note as above - can we check the literal value of of this property vs that it is any string?

teallarson replied:
Yep!

evantahler commented on __tests__/schemaGeneric.json:
```diff
@@ -0,0 +1,244 @@
+          "import": "./api/bin/import",
+          "watch": "tsc --watch"
+        }
+      ],
```
this big list of examples doesn't seem to gain us anything

evantahler commented on __tests__/schemaGeneric.json:
```diff
@@ -0,0 +1,244 @@
+    "dependencies": {
+      "$id": "#/properties/dependencies",
+      "type": "object",
+      "examples": [
```
this big list of examples doesn't seem to gain us anything

evantahler commented on __tests__/wordCheck:
```diff
@@ -30,13 +31,9 @@ for word in "${BAD_WORDS[@]}"; do
-    echo "$MATCHES"
-    echo "---------------"
-    exit 1
+    echo "$word: $MATCHES" 1>&2
```
Nice! Redirecting the echo from `stdout` to `stderr`.  This will get the promise in the jest test to reject, but it still common practice to explicitly `exit 1` to show the program has "failed" - please add that back.  This will also end the loop we are in early

evantahler commented on __tests__/packages.ts:
```diff
@@ -0,0 +1,119 @@
+          if (error || stderr) {
+            reject(error || stderr);
+          }
+          res();
```
Same as below - an early reject without a return will still also resolve later

krishnaglick replied:
I think doing `done(error || stderr)` and not wrapping all this in a promise may be the best way to handle it.

teallarson replied:
Went with the `done()` pattern for both the version and license checkers and piped out the most meaningful messaging for each checker (license checker shows it in `stderr`, but the version checker shows it in `stdout`)

evantahler reviewed (changes requested):
Good setup! I like the consistency of having a top-level test suite, rather then some random bash scripts we run.

I think there are a few topics to look into:
1. Don't import/require anything from grouparoo/core or grouparoo/spec-helper.  They shouldn't be needed and I'm worried that will have an adverse effect on the publishing pipeline
2. Can JSON-schema assert the contents of the JSON rather than the types?  For example, we don't really care that publish config is /any/ string in the packages, we really do want to assert that it is "public".  I think that's a better reflection of the goal.

Commit:
cleanup pt 1

grouparoo-bot committed:
Updating things automatically

Commit:
version checker good, license checker wip

Commit:
tsc cleanup, meaningful errors on test failure

Commit:
fix ci job template

teallarson commented:
Changes since last review:
- Tighter assertions for the `publishConfig` (Implemented by adding a third schema for `restricted` access configs and switching some logic for which schema to use.  The only packages using the `schemaGeneric` now are ones that do not require a `publishConfig`.)
- Tighter assertion for `engine`
- TSConfig and CI Job cleanup
- Tests for version and license checking fail correctly now, and display helpful errors (ie: what package it failed on and why)

Commit:
node engine schema pattern

Commit:
typo fix

grouparoo-bot committed:
Updating things automatically

evantahler commented on __tests__/packages.ts:
```diff
@@ -0,0 +1,120 @@
+
+  const pkgFile =
+    "/Users/teal/workspace/grouparoo/grouparoo/plugins/@grouparoo/customerio/package.json";
+  describe.each(packageFiles)(`%s`, (pkgFile) => {
```
I would not have expected this to pass on CI - there's no `/teal` folder there!  This should be fixed, but also... why did it pass?

teallarson replied:
Ah.  This was left over from debugging, oops.... it's still passing because the `pkgFile` referenced below is from the loop, not the one declared up here.

evantahler replied:
Ah! Then 👍 once this is removed!

Commit:
cleanup

Commit:
fix docker section

grouparoo-bot committed:
Updating things automatically
